test/decl/typealias/typealias.swift

```diff
@@ -122,3 +121,24 @@ extension A<Float,Int> {}  // expected-error {{constrained extension must be dec
 extension C<T> {}  // expected-error {{use of undeclared type 'T'}}
 extension C<Int> {}  // expected-error {{constrained extension must be declared on the unspecialized generic type 'MyType' with constraints specified by a 'where' clause}}
 
+
+// Allow typealias inside protocol, but don't allow it in where clauses (at least not yet)
+protocol Col {
+  associatedtype Elem
+  var elem: Elem { get }
+}
+
+protocol CB {
+  associatedtype C : Col
+  typealias E = C.Elem
+  
+  func setIt(element: E)
+}
+
+// expected-error @+1 {{'CB' typealias 'E' can not be used as a generic constraint; use associatedtype instead}}
+func go1<T : CB, U : Col where U.Elem == T.E>(col: U, builder: T) { // expected-error {{'E' is not a member type of 'T'}}
```

Comment:
Yes, it's hopefully just temporary - I'm trying to break these changes into smaller pieces. This pull avoids touching ArchetypeBuilder except for the temporary diagnosis with the limitation.  I have some more work there (stashed at the moment...) that makes generic constraints work for straightforward alias cases (like the `typealias Element = Iterator.Element` that you mentioned), but it isn't quite correct yet. It'll be next up!